### PR TITLE
US-2.8: Joining events

### LINKED ISSUE #66: US-2.8: Joining Events
**User Story Description**

As a user, I want to be able to join an event that interests me.

**User Acceptance Flow**

1. User logs in (Logged in user can be hard coded)
1. User finds an event that interests them on the Discover tab of the Home page
2.a User presses on the event card, bringing us to the preview page for the event. User presses "Join" on the event preview page
2.b User presses the join event button on the event card

**Acceptance Criteria**

- [ ] GraphQL mutation for adding users as participants to an event is implemented
- [ ] User Acceptance Flow is respected
- [ ] Documentation present where necessary
- [ ] Unit and UI tests are implemented for every component and function
- [ ] There must exist an automated Cypress acceptance test in GIF form
- [ ] Must match [Mockup](create-mockup-link)

*Dummy users can be used if logging is not completed - #99

### PR DESCRIPTION
**Pull request by leojrsilao**
This PR closes #66 

To properly test this PR you will need to set up your database such that User with id 1: 
- Is not involved in at least 1 event
 
**Steps to test joining events**
1. Make sure to run `npm i ` on the project to pickup the new dependencies 
2. Run `npm run gen-graphql` to generate the `graphql.tsx`
1. Run `nx serve mull-api` and `nx serve mull-ui`
2. Go to `http://localhost:4200/home/discover`
3. Join an event
3.a By clicking on the join event button on the event card
3.b By clicking on the event card, it brings you to the preview page for the event. Click on "Join" button on the event preview page 

**Steps to run the E2E tests:**
1. Make sure to run `npm i ` on the project to pickup the new dependencies 
1. Run the server `nx serve mull-api`
1. In the root directory run  `nx e2e mull-ui-e2e --watch` ( Cypress takes some time to open up)

### CODE REVIEW

File: apps/mull-api/src/app/event/event.service.ts [hunk 1]
```diff
@@ -47,33 +46,40 @@ export class EventService {
...
         new Brackets((qb) => {
-          qb.where('coHost.id != :userId', { userId }).orWhere('coHost.id IS NULL');
-        })
-      )
-      .andWhere(
-        new Brackets((qb) => {
-          qb.where('participant.id != :userId', { userId }).orWhere('participant.id IS NULL');
+          qb.where('coHost.id = :userId', { userId })
+            .orWhere('user.id = :userId', { userId })
+            .orWhere('host.id = :userId', { userId });
```
Comment (osahtout) [inline apps/mull-api/src/app/event/event.service.ts]: I'm a little confused, how come we're checking if the host has joined the event, or am I miss interpreting something?

File: apps/mull-api/src/app/event/event.service.ts [hunk 2]
```diff
@@ -47,33 +46,40 @@ export class EventService {
...
+   * Find all the events that the user has not joined, created or is not co-hosting
+   * @param userId
+   */
+  async findDiscoverEvent(userId: number): Promise<Event[]> {
+    const discoverableEvents = await this.eventRepository.query(
+      `
+    SELECT DISTINCT id, title, startDate, endDate, description, imageId, hostId, locationId, restriction FROM event LEFT JOIN event_participants ON event.id = event_participants.eventId
+    WHERE id NOT IN (SELECT event.id FROM event WHERE event.hostId = ${userId}) AND 
+    id NOT IN (SELECT event_participants.eventId FROM event_participants WHERE event_participants.userId = ${userId}) AND 
+    id NOT IN (SELECT event_cohosts.eventId FROM event_cohosts WHERE event_cohosts.userId = ${userId})`
```
Comment (osahtout) [inline apps/mull-api/src/app/event/event.service.ts:82]: This is just for curiosity to know if this way is better

Comment (cristian-aldea) [reply apps/mull-api/src/app/event/event.service.ts:82]: Yeah writing raw SQL isn't very appealing imo. What was wrong with the previous approach?

Comment (leojrsilao) [reply apps/mull-api/src/app/event/event.service.ts:82]: In the previous approach, the discovered events data wouldn't be displayed correctly if there's more than one person that joined an event, there's most likely a way to do subqueries using query builder, but I'm not too familiar with that.

Comment (r-saba) [reply apps/mull-api/src/app/event/event.service.ts:82]: This could probably be translated to the ORM syntax, would just take some extra work. 
It was a complicated query which involved a lot of thinking, testing and debugging in workbench using SQL, so it was easier to just bring over that working code.

File: apps/mull-api/src/app/event/event.service.ts (hunk 1 above)
Comment (osahtout) [reply apps/mull-api/src/app/event/event.service.ts]: I think the event created by a user should be queried separately since it will belong in its own tab in the homepage and have unique logic such as cancelling events, upgrading users to coHosts and the ability to remove participants.

Comment (leojrsilao) [reply apps/mull-api/src/app/event/event.service.ts]: Right, the query shouldn't show the host, since it should show 'cancel event' instead for those users, will remove the host from this query. For coHost, would need to re-affirm with the team if they can leave or actually cancel the event as well

File: apps/mull-ui/src/app/pages/home/my-events/my-events.tsx [hunk 3]
```diff
@@ -16,7 +15,12 @@ export const MyEventsPage = ({ history }) => {
       ISerializedEvent
     >[];
     var eventCards = events.map((event, index) => (
-      <EventCard key={index} event={event} onClick={() => history.push(`/events/${event.id}`)} />
+      <EventCard
+        isJoined={true}
```
Comment (osahtout) [inline apps/mull-ui/src/app/pages/home/my-events/my-events.tsx:20]: Since this tab is where all the events that the user has created, I think it should have its own condition to display. We can keep this here for now and update this later in another ticket

File: apps/mull-ui-e2e/src/integration/US-2.8.spec.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      cy.get('#event-card-join').click();
+    });
+
+    it('should join an event on the preview event page', () => {
+      cy.visit('http://localhost:4200/home/discover');
+      cy.get(':nth-child(1) > .event-card-image').click();
+      cy.get('[data-testid=mull-button]').click();
+    });
+  });
+});
```
Comment (RGPosadas) [inline apps/mull-ui-e2e/src/integration/US-2.8.spec.ts:32]: I think you should also have e2e tests for leaving events. Otherwise, running the E2E test would always need 2 available events to join, which gets pretty annoying if you keep running it.

**Review by RGPosadas — CHANGES_REQUESTED** [review]
Make sure to update your snapshots as well!
Otherwise, functionality is good! We also need to create a new TASK for Omar's comment about events in `My Events`.

File: apps/mull-ui/src/generated/graphql.tsx [hunk 5]
```diff
@@ -83,8 +83,8 @@ export type Mutation = {
 
 
 export type MutationAddParticipantToEventArgs = {
-  eventId: Scalars['Float'];
-  userId: Scalars['Float'];
+  eventId: Scalars['Int'];
```
Comment (cristian-aldea) [inline apps/mull-ui/src/generated/graphql.tsx]: You left in your instructions to run `npm run gen-graphql`. You should run it and push the changes it brings

File: apps/mull-api/src/app/event/event.resolver.ts [hunk 6]
```diff
@@ -34,6 +34,14 @@ export class EventResolver {
     return this.eventService.findJoinedEvents(id);
   }
 
+  @Query(/* istanbul ignore next */ () => Boolean)
+  async participatingEvent(
+    @Args('eventId', { type: /* istanbul ignore next */ () => Int }) eventId: number,
+    @Args('userId', { type: /* istanbul ignore next */ () => Int }) userId: number
+  ) {
+    return !!(await this.eventService.findJoinedEvent(eventId, userId));
```
Comment (cristian-aldea) [inline apps/mull-api/src/app/event/event.resolver.ts]: `findJoinedEvent` returns a promise for an event. Instead of doing the `!!` thing you can simply change what `findJoinedEvent` returns to be a boolean. when successful

Comment (cristian-aldea) [inline apps/mull-api/src/app/event/event.resolver.ts]: The naming of these methods is super confusing. Also, try to follow the general naming you see in #147 .

The resolver method could be called `isParticipant`, and the service name could be `isUserAttendingEvent`

File: apps/mull-ui/src/graphql/mutations.gql [hunk 7]
```diff
@@ -19,3 +19,13 @@ mutation CreateUser($createUserInput: CreateUserInput!) {
...
   }
 }
+
+# User joins an event
+mutation JoinEvent($eventId: Int!, $userId: Int!) {
+  addParticipantToEvent(eventId: $eventId, userId: $userId)
+}
+
+# User leaves an event
+mutation RemoveJoinedEvent($eventId: Int!, $userId: Int!) {
```
Comment (cristian-aldea) [inline apps/mull-ui/src/graphql/mutations.gql]:
```suggestion
mutation LeaveEvent($eventId: Int!, $userId: Int!) {
```

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.scss [hunk 8]
```diff
@@ -16,6 +16,12 @@
   margin: 0.9rem 0;
 }
 
+.event-page-joined-button {
+  border: 1.5px solid #27b09a;
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.scss]: Use the colour variable

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.tsx [hunk 9]
```diff
@@ -77,9 +88,31 @@ export const EventPageInfo = ({
...
+          onClick={(e) => {
+            e.stopPropagation();
+            setJoined(!joined);
+            joinEvent({ variables: { eventId, userId } });
+          }}
+          type={buttonType}
+        >
+          {isReview ? 'Create' : 'Join'}
+        </MullButton>
+      )}
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.tsx]: Most of the code is being duplicated. You can simplify this by only putting a conditional on the differences. for example:

```tsx
      <MullButton
        className={`event-page-button ${joined ? 'event-page-joined-button' : ''}`}
        onClick={(e) => {
          e.stopPropagation();
          setJoined(!joined);
          if (joined) {
            removeJoinedEvent({ variables: { eventId, userId } });
          } else {
            joinEvent({ variables: { eventId, userId } });
          }
        }}
        type={buttonType}
      >
        {joined ? 'Leave' : isReview ? 'Create' : 'Join'}
      </MullButton>
```

My solution might have some problems because I didn't test it, but you get the point

File: apps/mull-ui/src/graphql/queries.gql [hunk 10]
```diff
@@ -48,3 +48,21 @@ query GetParticipatingEvents($participatingEventsUserId: Int!) {
...
+}
+
+fragment EventContent on Event {
+  id
+  title
+  description
+  startDate
+  endDate
+  restriction
+}
```
Comment (cristian-aldea) [inline apps/mull-ui/src/graphql/queries.gql:73]: Why was this added? Just curious

**Review by cristian-aldea — COMMENTED** [review]
Great work so far, Everything works great and I don't have major complaints. However I do have some suggestions, so I'll leave them as comments.

File: apps/mull-api/src/app/event/event.service.ts (hunk 2 above)
Comment (cristian-aldea) [reply apps/mull-api/src/app/event/event.service.ts:82]: Oh yeah that's fine then, we can create a new ticket possibly to address this later.

File: apps/mull-ui/src/graphql/queries.gql (hunk 10 above)
Comment (leojrsilao) [reply apps/mull-ui/src/graphql/queries.gql:73]: It's not needed, but I think it would be cleaner if any new queries require the same event content, it can use this fragment

Comment (cristian-aldea) [reply apps/mull-ui/src/graphql/queries.gql:73]: You can leave it for now, but I think that kind of refactoring where you take out duplicates into a var or something should be done on a need basis.

File: apps/mull-ui/src/app/components/event-card/event-card.spec.tsx [hunk 11]
```diff
@@ -1,22 +1,46 @@
...
 import { render } from '@testing-library/react';
+import { createMemoryHistory } from 'history';
 import React from 'react';
+import { Router } from 'react-router-dom';
 import renderer from 'react-test-renderer';
 import { dummyEvent } from '../../../constants';
 import EventCard from './event-card';
 
 describe('EventCard', () => {
+  const history = createMemoryHistory();
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-card/event-card.spec.tsx]: Careful when declaring stuff outside of tests like this, since tests might start affecting each other.

File: apps/mull-ui/src/app/components/event-card/event-card.spec.tsx [hunk 12]
```diff
@@ -1,22 +1,46 @@
...
   it('should render successfully', () => {
-    const { baseElement } = render(<EventCard event={dummyEvent} />);
+    const { baseElement } = render(
+      <MockedProvider>
+        <Router history={history}>
+          <EventCard event={dummyEvent} />
+        </Router>
+      </MockedProvider>
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-card/event-card.spec.tsx:17]: You could put this to a separate function if you want, see `renderHelper` in `login.spec.tsx` ([link](https://github.com/RGPosadas/Mull/blob/develop/apps/mull-ui/src/app/pages/login/login.spec.tsx#L12)). The history can be a parameter, or in your case, it could be created inside the helper function

File: apps/mull-api/src/app/event/event.service.ts [hunk 13]
```diff
@@ -47,33 +46,42 @@ export class EventService {
...
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Find all the events that the user has not joined, created or is not co-hosting
+   * @param userId
+   */
+  async findDiscoverEvent(userId: number): Promise<Event[]> {
```
Comment (RGPosadas) [inline apps/mull-api/src/app/event/event.service.ts]:
```suggestion
  async findDiscoverableEvents(userId: number): Promise<Event[]> {
```
Since you're returning an array of events (and also to match the returned variable)

File: apps/mull-ui/src/app/components/event-card/event-card.tsx [hunk 14]
```diff
@@ -3,20 +3,29 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
...
 }
 
-export const EventCard = ({ event, style = {}, onClick }: EventCardProps) => {
-  // TODO; set joined based on if current user is part of event
-  const [joined, setJoined] = useState<boolean>(false);
+export const EventCard = ({ event, style = {}, onClick, isJoined = false }: EventCardProps) => {
   const { day, month, time } = formatDate(new Date(event.startDate));
+
+  const [joined, setJoined] = useState<boolean>(isJoined);
+  const eventId = Number(event.id);
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-card/event-card.tsx]: Isn't `event.id` already a number? If it's not. Then the TypeScript type annotation of ISerializedEvent is wrong and needs to be tweaked

EDIT: just confirmed that it's a string, ISerializedEvent needs to be updated so that this is reflected.

Not gonna block over this. If you don't address this in this PR then please make a ticket for it

**Review by RGPosadas — COMMENTED** [review]
Nice work! Had a smol comment but not a blocker.

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.spec.tsx [hunk 15]
```diff
@@ -1,17 +1,35 @@
+import { MockedProvider } from '@apollo/client/testing';
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/event-page-info/event-page-info.spec.tsx:1]: Same comments as for `apps/mull-ui/src/app/components/event-card/event-card.spec.tsx`

**Review by cristian-aldea — APPROVED** [review]
Great Work! Thanks for addressing my comments. Had some more concerns but they aren't pressing. Please reply to them thou

- [x] GraphQL mutation for adding users as participants to an event is implemented
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test in GIF form
- [x] Must match [Mockup](create-mockup-link)

File: apps/mull-api/src/app/event/event.service.ts (hunk 13 above)
Comment (RGPosadas) [reply apps/mull-api/src/app/event/event.service.ts]: You can ignore this and we can stick [to this method name](https://github.com/RGPosadas/Mull/pull/147/files#diff-b31268315353d230075823e831673a7cc40d3c83bf1720d4b5c2b2689131e4daL54)

Comment (cristian-aldea) [reply apps/mull-api/src/app/event/event.service.ts]: Have a look at the rename I did in #147, lemme know if you don't like it or have another preference

**Review by RGPosadas — APPROVED** [review]
Good work! LGTM 🚀 

**Acceptance Criteria**

- [x] GraphQL mutation for adding users as participants to an event is implemented
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test in GIF form
- [x] Must match [Mockup](create-mockup-link)

File: apps/mull-ui/src/app/components/event-card/event-card.tsx (hunk 14 above)
Comment (leojrsilao) [reply apps/mull-ui/src/app/components/event-card/event-card.tsx:22]: I will create a ticket for it

Comment (leojrsilao) [reply apps/mull-ui/src/app/components/event-card/event-card.tsx]: Addressed

File: apps/mull-ui/src/app/components/event-card/event-card.spec.tsx (hunk 12 above)
Comment (leojrsilao) [reply apps/mull-ui/src/app/components/event-card/event-card.spec.tsx:17]: Addressed

File: apps/mull-ui/src/app/components/event-page-info/event-page-info.spec.tsx (hunk 15 above)
Comment (leojrsilao) [reply apps/mull-ui/src/app/components/event-page-info/event-page-info.spec.tsx:1]: Addressed
